# Machine Learning - Continuous Assessment - Class Group: TU060 <a class="tocSkip">

# MSc in Computer Science - Data Science (Part Time) <a class="tocSkip">

# Student Name: Ciaran Finnegan  <a class="tocSkip">

# Student No: D21124026 <a class="tocSkip">

#### Student EMail: D21124026@mytudublin.ie <a class="tocSkip">

## May 2022<a class="tocSkip">

# 

# H&M Personalised Fashion Recommendations <a class="tocSkip">


#### This project takes the datasets provided by Kaggle for the H&M Personalised Fashion Recommendations competition.  <a class="tocSkip">

Ipso forum...

## <a class="tocSkip">

#### The CA follows a n-part structure. <a class="tocSkip">

Section 1: Ipso forum...

Section 2: Ipso forum...

Section 3: Ipso forum...


### <a class="tocSkip">

<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Jupyter-Notebook-Set-Up" data-toc-modified-id="Jupyter-Notebook-Set-Up-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Jupyter Notebook Set Up</a></span><ul class="toc-item"><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Import Libraries</a></span><ul class="toc-item"><li><span><a href="#General-Utility-Libraries" data-toc-modified-id="General-Utility-Libraries-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>General Utility Libraries</a></span></li><li><span><a href="#Panda-+-Numpy" data-toc-modified-id="Panda-+-Numpy-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Panda + Numpy</a></span></li><li><span><a href="#Matplotlib-Library" data-toc-modified-id="Matplotlib-Library-1.1.3"><span class="toc-item-num">1.1.3&nbsp;&nbsp;</span>Matplotlib Library</a></span></li><li><span><a href="#Plotly-Libraries" data-toc-modified-id="Plotly-Libraries-1.1.4"><span class="toc-item-num">1.1.4&nbsp;&nbsp;</span>Plotly Libraries</a></span></li></ul></li></ul></li><li><span><a href="#Project-Data-Import,-Analysis-and-Preparation" data-toc-modified-id="Project-Data-Import,-Analysis-and-Preparation-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Project Data Import, Analysis and Preparation</a></span><ul class="toc-item"><li><span><a href="#Load-and-Sample-Kaggle-H&amp;M-Datasets" data-toc-modified-id="Load-and-Sample-Kaggle-H&amp;M-Datasets-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Load and Sample Kaggle H&amp;M Datasets</a></span><ul class="toc-item"><li><span><a href="#Set-Up-File-Paths" data-toc-modified-id="Set-Up-File-Paths-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Set Up File Paths</a></span></li><li><span><a href="#H&amp;M-Transaction-Dataset" data-toc-modified-id="H&amp;M-Transaction-Dataset-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>H&amp;M Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Articles-Dataset" data-toc-modified-id="H&amp;M-Articles-Dataset-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>H&amp;M Articles Dataset</a></span></li><li><span><a href="#H&amp;M-Customers-Dataset" data-toc-modified-id="H&amp;M-Customers-Dataset-2.1.4"><span class="toc-item-num">2.1.4&nbsp;&nbsp;</span>H&amp;M Customers Dataset</a></span></li></ul></li><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#H&amp;M-Sampled-Transaction-Dataset" data-toc-modified-id="H&amp;M-Sampled-Transaction-Dataset-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>H&amp;M Sampled Transaction Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Customer-Dataset" data-toc-modified-id="H&amp;M-Sampled-Customer-Dataset-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>H&amp;M Sampled Customer Dataset</a></span></li><li><span><a href="#H&amp;M-Sampled-Articles-Dataset" data-toc-modified-id="H&amp;M-Sampled-Articles-Dataset-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>H&amp;M Sampled Articles Dataset</a></span></li></ul></li><li><span><a href="#Data-Preparation-and-Feature-Enrichment" data-toc-modified-id="Data-Preparation-and-Feature-Enrichment-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Data Preparation and Feature Enrichment</a></span><ul class="toc-item"><li><span><a href="#Transaction-Data-Feature-Enrichment-(Sample)" data-toc-modified-id="Transaction-Data-Feature-Enrichment-(Sample)-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Transaction Data Feature Enrichment (Sample)</a></span></li><li><span><a href="#Extend-Dataframe-with-Customer-Data" data-toc-modified-id="Extend-Dataframe-with-Customer-Data-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>Extend Dataframe with Customer Data</a></span></li><li><span><a href="#Extend-Dataframe-with-Articles-Data" data-toc-modified-id="Extend-Dataframe-with-Articles-Data-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>Extend Dataframe with Articles Data</a></span></li><li><span><a href="#Final-Data-Preparation-Steps" data-toc-modified-id="Final-Data-Preparation-Steps-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>Final Data Preparation Steps</a></span></li></ul></li></ul></li><li><span><a href="#Split-Data:-Train-and-Test-Sets" data-toc-modified-id="Split-Data:-Train-and-Test-Sets-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Split Data: Train and Test Sets</a></span></li><li><span><a href="#Model-Preparation" data-toc-modified-id="Model-Preparation-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Model Preparation</a></span><ul class="toc-item"><li><span><a href="#Set-Up-the-Scaling-Algorithm" data-toc-modified-id="Set-Up-the-Scaling-Algorithm-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Set Up the Scaling Algorithm</a></span></li><li><span><a href="#Set-Up-the-Columns-to-be-Scaled" data-toc-modified-id="Set-Up-the-Columns-to-be-Scaled-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Set Up the Columns to be Scaled</a></span></li><li><span><a href="#Scale-the-Training-Dataset" data-toc-modified-id="Scale-the-Training-Dataset-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Scale the Training Dataset</a></span></li></ul></li><li><span><a href="#K-Nearest-Neighbour-Operations" data-toc-modified-id="K-Nearest-Neighbour-Operations-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>K Nearest Neighbour Operations</a></span><ul class="toc-item"><li><span><a href="#Create-Functions-for-NN-Analysis" data-toc-modified-id="Create-Functions-for-NN-Analysis-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Create Functions for NN Analysis</a></span><ul class="toc-item"><li><span><a href="#Set-up-function-to-Build-Nearest-Neighbours-Model" data-toc-modified-id="Set-up-function-to-Build-Nearest-Neighbours-Model-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>Set up function to Build Nearest Neighbours Model</a></span></li><li><span><a href="#Set-up-function-to--Nearest-Neighbours-(+-Articles-bought)-for-a-Customer" data-toc-modified-id="Set-up-function-to--Nearest-Neighbours-(+-Articles-bought)-for-a-Customer-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>Set up function to  Nearest Neighbours (+ Articles bought) for a Customer</a></span></li><li><span><a href="#Find-Top-Selling-Products-in-'Neighbourhood'-for-a-Customer" data-toc-modified-id="Find-Top-Selling-Products-in-'Neighbourhood'-for-a-Customer-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>Find Top Selling Products in 'Neighbourhood' for a Customer</a></span></li><li><span><a href="#Invoke-Function-to-Build-KNN-Model" data-toc-modified-id="Invoke-Function-to-Build-KNN-Model-5.1.4"><span class="toc-item-num">5.1.4&nbsp;&nbsp;</span>Invoke Function to Build KNN Model</a></span></li></ul></li></ul></li><li><span><a href="#Evaluate-Model" data-toc-modified-id="Evaluate-Model-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Evaluate Model</a></span><ul class="toc-item"><li><span><a href="#Define-Functions-to-Extract-Customer-IDs-and-Article-Predictions-from-Test-Data" data-toc-modified-id="Define-Functions-to-Extract-Customer-IDs-and-Article-Predictions-from-Test-Data-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Define Functions to Extract Customer IDs and Article Predictions from Test Data</a></span><ul class="toc-item"><li><span><a href="#Define-Function-to-read-Customer-IDs-from-Submission-File" data-toc-modified-id="Define-Function-to-read-Customer-IDs-from-Submission-File-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>Define Function to read Customer IDs from Submission File</a></span></li><li><span><a href="#Define-Function-to-Obtain-Customer-Transactions" data-toc-modified-id="Define-Function-to-Obtain-Customer-Transactions-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>Define Function to Obtain Customer Transactions</a></span></li><li><span><a href="#Define-Function-to-Obtain-Customer-Features" data-toc-modified-id="Define-Function-to-Obtain-Customer-Features-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>Define Function to Obtain Customer Features</a></span></li><li><span><a href="#Define-Function-to-Obtain-Article-Features" data-toc-modified-id="Define-Function-to-Obtain-Article-Features-6.1.4"><span class="toc-item-num">6.1.4&nbsp;&nbsp;</span>Define Function to Obtain Article Features</a></span></li><li><span><a href="#Define-Function-to-Obtain-Article-Predictions" data-toc-modified-id="Define-Function-to-Obtain-Article-Predictions-6.1.5"><span class="toc-item-num">6.1.5&nbsp;&nbsp;</span>Define Function to Obtain Article Predictions</a></span></li><li><span><a href="#Define-Functions-to-Parse-Article-Predictions-for-Kaggle-Submission" data-toc-modified-id="Define-Functions-to-Parse-Article-Predictions-for-Kaggle-Submission-6.1.6"><span class="toc-item-num">6.1.6&nbsp;&nbsp;</span>Define Functions to Parse Article Predictions for Kaggle Submission</a></span></li></ul></li><li><span><a href="#Define-Functions-to-Evaluate-Model-Predictions-aginst-Test-Data" data-toc-modified-id="Define-Functions-to-Evaluate-Model-Predictions-aginst-Test-Data-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Define Functions to Evaluate Model Predictions aginst Test Data</a></span><ul class="toc-item"><li><span><a href="#Set-up-function-to-return-latest-Customer-transactions-in-the-Test-Data" data-toc-modified-id="Set-up-function-to-return-latest-Customer-transactions-in-the-Test-Data-6.2.1"><span class="toc-item-num">6.2.1&nbsp;&nbsp;</span>Set up function to return latest Customer transactions in the Test Data</a></span></li><li><span><a href="#Define-Function-to-Return-Evaluation-Score" data-toc-modified-id="Define-Function-to-Return-Evaluation-Score-6.2.2"><span class="toc-item-num">6.2.2&nbsp;&nbsp;</span>Define Function to Return Evaluation Score</a></span></li><li><span><a href="#Iterate-Through-Test-Data-and-Generate-Predictions" data-toc-modified-id="Iterate-Through-Test-Data-and-Generate-Predictions-6.2.3"><span class="toc-item-num">6.2.3&nbsp;&nbsp;</span>Iterate Through Test Data and Generate Predictions</a></span></li></ul></li></ul></li><li><span><a href="#Build-Submission" data-toc-modified-id="Build-Submission-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Build Submission</a></span><ul class="toc-item"><li><span><a href="#Read-Sample-Submission-File" data-toc-modified-id="Read-Sample-Submission-File-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Read Sample Submission File</a></span></li><li><span><a href="#Iterate-Through-Submission-File-to-Generate-Article-Predictions" data-toc-modified-id="Iterate-Through-Submission-File-to-Generate-Article-Predictions-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Iterate Through Submission File to Generate Article Predictions</a></span></li><li><span><a href="#Write-Predictions-Submission-File-to-CSV" data-toc-modified-id="Write-Predictions-Submission-File-to-CSV-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>Write Predictions Submission File to CSV</a></span><ul class="toc-item"><li><span><a href="#SetUp-Disk-Directory-for-Submission-File" data-toc-modified-id="SetUp-Disk-Directory-for-Submission-File-7.3.1"><span class="toc-item-num">7.3.1&nbsp;&nbsp;</span>SetUp Disk Directory for Submission File</a></span></li><li><span><a href="#Create-Submission-.CSV-File" data-toc-modified-id="Create-Submission-.CSV-File-7.3.2"><span class="toc-item-num">7.3.2&nbsp;&nbsp;</span>Create Submission .CSV File</a></span></li></ul></li></ul></li><li><span><a href="#References" data-toc-modified-id="References-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>References</a></span></li></ul></div>

## Jupyter Notebook Set Up

### Import Libraries

Note: It is assumed that the matplotlib and plotly library has been installed on the user's machine before it is available for import.

```bash
> pip install matplotlib
> pip install plotly
```


#### General Utility Libraries

In [1]:
import os

#### Panda + Numpy

In [2]:
# Python libraries for data manipulation
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import NearestNeighbors

Set up additional libraries for graphical presentations. A number of earlier graphs use Matplotlib. In later sections, Plotly graphs are used.

#### Matplotlib Library

In [3]:
#import matplotlib.pyplot as plt

#### Plotly Libraries

In [4]:
# For interactive graphics and global displays
#import plotly.express as px
#import plotly.graph_objects as go

### <a class="tocSkip">

In [5]:
import warnings
warnings.filterwarnings("ignore")

### <a class="tocSkip">

## Project Data Import, Analysis and Preparation

### Load and Sample Kaggle H&M Datasets

### The assignment starts with a Kaggle H&M Personalised Fashion Recommendations datasets. <a class="tocSkip">

#### Set Up File Paths 

In [6]:
# Set up a Global path file variable to use when loading each dataset
Path_To_Inputs = '../input'

#### H&M Transaction Dataset 

In [7]:
# Load H&M Transaction dataset
df_trxns = pd.read_csv(os.path.join(Path_To_Inputs,'transactions_train.csv'))

In [8]:
# Display Number of Rows, Number of Columns in Transaction dataset prior to Sampling
df_trxns.shape

(31788324, 5)

In [9]:
# Extract random sample of Transactions (without replacement) - 0.1% of records
df_trxns_sample = df_trxns.sample(frac=0.20, replace=False)

In [10]:
# Display Number of Rows, Number of Columns in Transaction dataset after Sampling
df_trxns_sample.shape

(6357665, 5)

In [11]:
# Check out the index
df_trxns_sample.head()

t_dat                                        customer_id  \
18031857  2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   
8577547   2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   
12171604  2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   
13137712  2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   
9419682   2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   

          article_id     price  sales_channel_id  
18031857   733099003  0.010153                 1  
8577547    606395014  0.059305                 2  
12171604   665963007  0.006085                 1  
13137712   702890001  0.010153                 2  
9419682    375585027  0.007898                 2

#### H&M Articles Dataset 

In [12]:
# Read file, located in 'input' folder above this notebook, for articles data - containing meta data on purchasable items.
df_Articles = pd.read_csv(os.path.join(Path_To_Inputs,'articles.csv'))

In [13]:
# Display Number of Rows, Number of Columns in Articles dataset
df_Articles.shape

(105542, 25)

In [14]:
# Extract only the H&M articles contained in the sample Transaction dataset 
articleIDs_in_Trxn_Sample  = set(df_trxns_sample["article_id"])

In [15]:
# Create dataset that only contains H&M articles in the sample Transaction dataset
df_articles_sample = df_Articles[df_Articles["article_id"].isin(articleIDs_in_Trxn_Sample)]

In [16]:
# Display Number of Rows, Number of Columns in Sample Articles dataset
df_articles_sample.shape

(94127, 25)

#### H&M Customers Dataset 

In [17]:
# Load H&M Customer dataset
df_customers = pd.read_csv(os.path.join(Path_To_Inputs,'customers.csv'))

In [18]:
# Display Number of Rows, Number of Columns in Customers dataset prior to Sampling
df_customers.shape

(1371980, 7)

In [19]:
# Extract only the H&M Customers contained in the sample Transaction dataset 
custIDs_in_Trxn_Sample  = set(df_trxns_sample["customer_id"])

In [20]:
#Count Customer Ids in Transaction Sample
len(custIDs_in_Trxn_Sample)

1020354

In [21]:
# Create dataset that only contains H&M Customers from the sample Transaction dataset
df_customers_sample = df_customers[df_customers["customer_id"].isin(custIDs_in_Trxn_Sample)]

In [22]:
# Display Number of Rows, Number of Columns in Sample Customers dataset
df_customers_sample.shape

(1020354, 7)

###### 

### Exploratory Data Analysis

### Display key data attributes and distributions in Sampled H&M datasets. <a class="tocSkip">

#### H&M Sampled Transaction Dataset 

In [23]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - after Sampling of original dataset
df_trxns_sample.head()

t_dat                                        customer_id  \
18031857  2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   
8577547   2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   
12171604  2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   
13137712  2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   
9419682   2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   

          article_id     price  sales_channel_id  
18031857   733099003  0.010153                 1  
8577547    606395014  0.059305                 2  
12171604   665963007  0.006085                 1  
13137712   702890001  0.010153                 2  
9419682    375585027  0.007898                 2

In [24]:
# Display datatype of colums in original H&M Transaction dataset
df_trxns_sample.dtypes

t_dat                object
customer_id          object
article_id            int64
price               float64
sales_channel_id      int64
dtype: object

#### H&M Sampled Customer Dataset 

In [25]:
# Display basic column information for the initial Kaggle 'customers' datatset - after Sampling of original dataset
df_customers_sample.head()

customer_id   FN  Active  \
0  00000dbacae5abe5e23885899a1fa44253a17956c6d1c3...  NaN     NaN   
1  0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...  NaN     NaN   
2  000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...  NaN     NaN   
4  00006413d8573cd20ed7128e53b7b13819fe5cfc2d801f...  1.0     1.0   
5  000064249685c11552da43ef22a5030f35a147f723d5b0...  NaN     NaN   

  club_member_status fashion_news_frequency   age  \
0             ACTIVE                   NONE  49.0   
1             ACTIVE                   NONE  25.0   
2             ACTIVE                   NONE  24.0   
4             ACTIVE              Regularly  52.0   
5                NaN                    NaN   NaN   

                                         postal_code  
0  52043ee2162cf5aa7ee79974281641c6f11a68d276429a...  
1  2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...  
2  64f17e6a330a85798e4998f62d0930d14db8db1c054af6...  
4  25fa5ddee9aac01b35208d01736e57942317d756b32ddd...  
5  2c29ae653a9282cce4151bd87643c907644e09541abc28...

In [26]:
df_customers_sample.dtypes

customer_id                object
FN                        float64
Active                    float64
club_member_status         object
fashion_news_frequency     object
age                       float64
postal_code                object
dtype: object

#### H&M Sampled Articles Dataset 

In [27]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code          prod_name  product_type_no  \
0   108775015        108775          Strap top              253   
1   108775044        108775          Strap top              253   
2   108775051        108775      Strap top (1)              253   
3   110065001        110065  OP T-shirt (Idro)              306   
4   110065002        110065  OP T-shirt (Idro)              306   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
2          Vest top  Garment Upper body                  1010017   
3               Bra           Underwear                  1010016   
4               Bra           Underwear                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
2                    Stripe                 11         Off White  ...   
3                     Solid                  9             Black  ...   
4                     Solid                 10             White  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
2     Jersey Basic          A        Ladieswear              1   
3   Clean Lingerie          B  Lingeries/Tights              1   
4   Clean Lingerie          B  Lingeries/Tights              1   

   index_group_name section_no            section_name garment_group_no  \
0        Ladieswear         16  Womens Everyday Basics             1002   
1        Ladieswear         16  Womens Everyday Basics             1002   
2        Ladieswear         16  Womens Everyday Basics             1002   
3        Ladieswear         61         Womens Lingerie             1017   
4        Ladieswear         61         Womens Lingerie             1017   

   garment_group_name                                        detail_desc  
0        Jersey Basic            Jersey top with narrow shoulder straps.  
1        Jersey Basic            Jersey top with narrow shoulder straps.  
2        Jersey Basic            Jersey top with narrow shoulder straps.  
3   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  
4   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  

[5 rows x 25 columns]

In [28]:
df_articles_sample.dtypes

article_id                       int64
product_code                     int64
prod_name                       object
product_type_no                  int64
product_type_name               object
product_group_name              object
graphical_appearance_no          int64
graphical_appearance_name       object
colour_group_code                int64
colour_group_name               object
perceived_colour_value_id        int64
perceived_colour_value_name     object
perceived_colour_master_id       int64
perceived_colour_master_name    object
department_no                    int64
department_name                 object
index_code                      object
index_name                      object
index_group_no                   int64
index_group_name                object
section_no                       int64
section_name                    object
garment_group_no                 int64
garment_group_name              object
detail_desc                     object
dtype: object

###### 

### Data Preparation and Feature Enrichment

### Construct a dataset upon which to build ML models. <a class="tocSkip">

### Begin with Transaction Dataset (sampled) and merge in customer and article data. <a class="tocSkip">

#### Transaction Data Feature Enrichment (Sample)

In [29]:
# Create new dataset for modelling. Initial basis for new dataset is the sampled transaction dataset.
df_HM_CustTrxn = df_trxns_sample

In [30]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  \
18031857  2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   
8577547   2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   
12171604  2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   
13137712  2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   
9419682   2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   

          article_id     price  sales_channel_id  
18031857   733099003  0.010153                 1  
8577547    606395014  0.059305                 2  
12171604   665963007  0.006085                 1  
13137712   702890001  0.010153                 2  
9419682    375585027  0.007898                 2

In [31]:
# Decompose Datetime information 
df_HM_CustTrxn['t_dat'] = pd.to_datetime(df_HM_CustTrxn['t_dat'], format="%Y-%m-%d")

# Add Column for Transaction Year and convert to integer value
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['t_dat'].dt.strftime('%Y')
df_HM_CustTrxn['txn_year'] = df_HM_CustTrxn['txn_year'].astype(str).astype(int)

# Add Column for Transaction Month and convert to integer value
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['t_dat'].dt.strftime('%m')
df_HM_CustTrxn['txn_mth'] = df_HM_CustTrxn['txn_mth'].astype(str).astype(int)

# Add Column for Transaction Day and convert to integer value
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['t_dat'].dt.strftime('%d')
df_HM_CustTrxn['txn_day'] = df_HM_CustTrxn['txn_day'].astype(str).astype(int)

In [32]:
# Filter out column to drop timestamp attribute
# df_HM_CustTrxn = df_HM_CustTrxn.iloc[:,1:8]

In [33]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  \
18031857 2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   
8577547  2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   
12171604 2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   
13137712 2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   
9419682  2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   

          article_id     price  sales_channel_id  txn_year  txn_mth  txn_day  
18031857   733099003  0.010153                 1      2019       10       18  
8577547    606395014  0.059305                 2      2019        4       10  
12171604   665963007  0.006085                 1      2019        6       20  
13137712   702890001  0.010153                 2      2019        7        2  
9419682    375585027  0.007898                 2      2019        4       27

In [34]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the Trxn (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the Trxn (sample) dataframe? 
 t_dat               False
customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
dtype: bool


[None]

#### Extend Dataframe with Customer Data

In [35]:
# Confirm dimensions of Trxn (sample) dataframe BEFORE merging with Customer data
df_HM_CustTrxn.shape

(6357665, 8)

#### Merge Customer Attributes into the Trxn (Sample) Dataframe. <a class="tocSkip">

In [36]:
# Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
df_HM_CustTrxn = pd.merge(df_HM_CustTrxn, df_customers_sample, how="left", on=["customer_id"])

In [37]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Customer data
df_HM_CustTrxn.shape

(6357665, 14)

In [38]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   733099003   
1 2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   606395014   
2 2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   665963007   
3 2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   702890001   
4 2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   375585027   

      price  sales_channel_id  txn_year  txn_mth  txn_day   FN  Active  \
0  0.010153                 1      2019       10       18  NaN     NaN   
1  0.059305                 2      2019        4       10  1.0     1.0   
2  0.006085                 1      2019        6       20  NaN     NaN   
3  0.010153                 2      2019        7        2  1.0     1.0   
4  0.007898                 2      2019        4       27  NaN     NaN   

  club_member_status fashion_news_frequency   age  \
0             ACTIVE                   NONE  27.0   
1             ACTIVE              Regularly  30.0   
2             ACTIVE                   NONE  41.0   
3             ACTIVE              Regularly  27.0   
4             ACTIVE                   NONE  39.0   

                                         postal_code  
0  f90755f75268c3bf1d9250b242a4c74d8e537e9e6796be...  
1  f122227d7ab3920e7877553f381ff5da244371c395f745...  
2  5ad9d934cb1824e6256941ca54160248ed17a3ce869d84...  
3  57108f94c0128a2ed727948bc3a8040caa87a57e8a933c...  
4  947dbca783c0ca13cd3edccd06c45ebc11e2a0f73e7d86...

In [39]:
# Filter out other Customer specific attributes apart from; age + fashion_news_frequency + club_member_status + postal code
# There are very many missing values in the other Customer dataset columns, so these will be ignored
#df_HM_CustTrxn = df_HM_CustTrxn.iloc[:, [0, 1, 2, 3, 4, 5, 6, 9, 10, 11, 12]]
df_HM_CustTrxn.drop(["FN", "Active"], axis=1, inplace=True)

In [40]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   733099003   
1 2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   606395014   
2 2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   665963007   
3 2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   702890001   
4 2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   375585027   

      price  sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0  0.010153                 1      2019       10       18             ACTIVE   
1  0.059305                 2      2019        4       10             ACTIVE   
2  0.006085                 1      2019        6       20             ACTIVE   
3  0.010153                 2      2019        7        2             ACTIVE   
4  0.007898                 2      2019        4       27             ACTIVE   

  fashion_news_frequency   age  \
0                   NONE  27.0   
1              Regularly  30.0   
2                   NONE  41.0   
3              Regularly  27.0   
4                   NONE  39.0   

                                         postal_code  
0  f90755f75268c3bf1d9250b242a4c74d8e537e9e6796be...  
1  f122227d7ab3920e7877553f381ff5da244371c395f745...  
2  5ad9d934cb1824e6256941ca54160248ed17a3ce869d84...  
3  57108f94c0128a2ed727948bc3a8040caa87a57e8a933c...  
4  947dbca783c0ca13cd3edccd06c45ebc11e2a0f73e7d86...

In [41]:
df_HM_CustTrxn.shape

(6357665, 12)

In [42]:
# Check for null values in the merged Trxn/Customer dataframe
[print('Any Null Values in the merged Trxn/Customer dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer dataframe? 
 t_dat                     False
customer_id               False
article_id                False
price                     False
sales_channel_id          False
txn_year                  False
txn_mth                   False
txn_day                   False
club_member_status         True
fashion_news_frequency     True
age                        True
postal_code               False
dtype: bool


[None]

#### Remove Rows with Missing Club Member Status Attributes. <a class="tocSkip">

In [43]:
# Check missing values in the club_member_status column  
df_null_club_member_status = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].isnull())]

[print('Numbers of rows with missing club member status?','\n' ,len(df_null_club_member_status.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing club member status attributes?\n{:06.3f}%'
                               .format((len(df_null_club_member_status.index)/len(df_HM_CustTrxn.index))*100))]


# For simplicity the rows with missing values in the club member status will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the club member status
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["club_member_status"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the club member status column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Numbers of rows with missing club member status? 
 12493

Precentage of total rows in Trxn dataframe with missing club member status attributes?
00.197%

Current Number of Rows/Columns in Trxn/Customer dataframe - with missing club mem status values removed 
 (6345172, 12)


[None]

#### Remove Rows with Missing Fashion News Frequency Attributes. <a class="tocSkip">

In [44]:
# Check missing values in the club_member_status column  
df_null_fn_freq = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].isnull())]

[print('Numbers of rows with missing fashion_news_frequency values?','\n' ,len(df_null_fn_freq.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?\n{:06.3f}%'
                               .format((len(df_null_fn_freq.index)/len(df_HM_CustTrxn.index))*100))]

# For simplicity the rows with missing values in the fashion_news_frequency column will be ignored
# Re-Build Txn/Customer Dataframe with no missing values in the fashion_news_frequency column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["fashion_news_frequency"].notnull())]

# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the fashion_news_frequency column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]

Numbers of rows with missing fashion_news_frequency values? 
 25896

Precentage of total rows in Trxn dataframe with missing fashion_news_frequency attributes?
00.408%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing Fash News Freq values removed 
 (6319276, 12)


[None]

#### Remove Rows with Missing Age Attributes. <a class="tocSkip">

In [45]:
# Check missing values in the age column  
df_null_age = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].isnull())]

[print('Numbers of rows with missing age values?','\n' ,len(df_null_age.index))]
[print('\nPrecentage of total rows in Trxn dataframe with missing age attributes?\n{:06.3f}%'
                               .format((len(df_null_age.index)/len(df_HM_CustTrxn.index))*100))]

# Re-Build Txn/Customer Dataframe with no missing values in the age column
df_HM_CustTrxn = df_HM_CustTrxn.loc[(df_HM_CustTrxn["age"].notnull())]


# Confirm dimensions of Trxn (sample) dataframe AFTER removing rows with missing values in the age column
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed','\n' ,df_HM_CustTrxn.shape)]

Numbers of rows with missing age values? 
 25150

Precentage of total rows in Trxn dataframe with missing age attributes?
00.398%

Current Number of Rows/Columns in Trxn/Customer dataframe - missing ages removed 
 (6294126, 12)


[None]

#### Encode the club_member_status + fashion_news_frequency attributes <a class="tocSkip">

In [46]:
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   733099003   
1 2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   606395014   
2 2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   665963007   
3 2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   702890001   
4 2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   375585027   

      price  sales_channel_id  txn_year  txn_mth  txn_day club_member_status  \
0  0.010153                 1      2019       10       18             ACTIVE   
1  0.059305                 2      2019        4       10             ACTIVE   
2  0.006085                 1      2019        6       20             ACTIVE   
3  0.010153                 2      2019        7        2             ACTIVE   
4  0.007898                 2      2019        4       27             ACTIVE   

  fashion_news_frequency   age  \
0                   NONE  27.0   
1              Regularly  30.0   
2                   NONE  41.0   
3              Regularly  27.0   
4                   NONE  39.0   

                                         postal_code  
0  f90755f75268c3bf1d9250b242a4c74d8e537e9e6796be...  
1  f122227d7ab3920e7877553f381ff5da244371c395f745...  
2  5ad9d934cb1824e6256941ca54160248ed17a3ce869d84...  
3  57108f94c0128a2ed727948bc3a8040caa87a57e8a933c...  
4  947dbca783c0ca13cd3edccd06c45ebc11e2a0f73e7d86...

In [47]:
def encode_Categor_Features(df, col_name):
       
    # The categorical values in dataframe must be encoded to numberical values
    df_encoded_cols = pd.get_dummies(df[col_name])
    df_encoded      = df.join(df_encoded_cols)
    
    # Remove the prior Categorical column from the Articles dataframe
    df_encoded.drop([col_name], axis=1, inplace=True)
    
    # Return the dataframe with encoded columsn added
    return df_encoded

In [48]:
cust_cols_to_encode = ["club_member_status","fashion_news_frequency"]
for column_name in cust_cols_to_encode:
    df_HM_CustTrxn = encode_Categor_Features(df_HM_CustTrxn, column_name)

In [49]:
# Drop Postal Code - a non numerical value that is not easily converted
df_HM_CustTrxn.drop(["postal_code"], axis=1, inplace=True)

In [50]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_HM_CustTrxn.head()

t_dat                                        customer_id  article_id  \
0 2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   733099003   
1 2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   606395014   
2 2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   665963007   
3 2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   702890001   
4 2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   375585027   

      price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  \
0  0.010153                 1      2019       10       18  27.0       1   
1  0.059305                 2      2019        4       10  30.0       1   
2  0.006085                 1      2019        6       20  41.0       1   
3  0.010153                 2      2019        7        2  27.0       1   
4  0.007898                 2      2019        4       27  39.0       1   

   LEFT CLUB  PRE-CREATE  Monthly  NONE  None  Regularly  
0          0           0        0     1     0          0  
1          0           0        0     0     0          1  
2          0           0        0     1     0          0  
3          0           0        0     0     0          1  
4          0           0        0     1     0          0

In [51]:
# Confirm dimensions of Trxn (sample) dataframe AFTER encoding Categorical Values
[print('\nCurrent Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed','\n' ,
                                                                                                   df_HM_CustTrxn.shape)]


Current Number of Rows/Columns in Trxn/Customer dataframe - after encoding of categorical values completed 
 (6294126, 16)


[None]

In [52]:
# Check for null values in the Transaction Dataframe
[print('Any Null Values in the merged Trxn/Customer (sample) dataframe?','\n' ,df_HM_CustTrxn.isnull().any())]

Any Null Values in the merged Trxn/Customer (sample) dataframe? 
 t_dat               False
customer_id         False
article_id          False
price               False
sales_channel_id    False
txn_year            False
txn_mth             False
txn_day             False
age                 False
ACTIVE              False
LEFT CLUB           False
PRE-CREATE          False
Monthly             False
NONE                False
None                False
Regularly           False
dtype: bool


[None]

#### Extend Dataframe with Articles Data

#### Remove 'Redundant' Categorical Columns in Articles (sample) Dataframe. <a class="tocSkip">

In [53]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code          prod_name  product_type_no  \
0   108775015        108775          Strap top              253   
1   108775044        108775          Strap top              253   
2   108775051        108775      Strap top (1)              253   
3   110065001        110065  OP T-shirt (Idro)              306   
4   110065002        110065  OP T-shirt (Idro)              306   

  product_type_name  product_group_name  graphical_appearance_no  \
0          Vest top  Garment Upper body                  1010016   
1          Vest top  Garment Upper body                  1010016   
2          Vest top  Garment Upper body                  1010017   
3               Bra           Underwear                  1010016   
4               Bra           Underwear                  1010016   

  graphical_appearance_name  colour_group_code colour_group_name  ...  \
0                     Solid                  9             Black  ...   
1                     Solid                 10             White  ...   
2                    Stripe                 11         Off White  ...   
3                     Solid                  9             Black  ...   
4                     Solid                 10             White  ...   

   department_name index_code        index_name index_group_no  \
0     Jersey Basic          A        Ladieswear              1   
1     Jersey Basic          A        Ladieswear              1   
2     Jersey Basic          A        Ladieswear              1   
3   Clean Lingerie          B  Lingeries/Tights              1   
4   Clean Lingerie          B  Lingeries/Tights              1   

   index_group_name section_no            section_name garment_group_no  \
0        Ladieswear         16  Womens Everyday Basics             1002   
1        Ladieswear         16  Womens Everyday Basics             1002   
2        Ladieswear         16  Womens Everyday Basics             1002   
3        Ladieswear         61         Womens Lingerie             1017   
4        Ladieswear         61         Womens Lingerie             1017   

   garment_group_name                                        detail_desc  
0        Jersey Basic            Jersey top with narrow shoulder straps.  
1        Jersey Basic            Jersey top with narrow shoulder straps.  
2        Jersey Basic            Jersey top with narrow shoulder straps.  
3   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  
4   Under-, Nightwear  Microfibre T-shirt bra with underwired, moulde...  

[5 rows x 25 columns]

In [54]:
# Remove 'redundant' categotical columns from Articles (sample) dataframe
# Consider a categorical column redundant when is 'partnered' with a numerical code 

redundo_catcols_to_drop =  ["prod_name", 
                            "product_type_name", 
                            "graphical_appearance_name",
                            "colour_group_name",
                            "perceived_colour_value_name",
                            "perceived_colour_master_name",
                            "department_name",
                            "index_name",
                            "index_group_name",
                            "section_name",
                            "garment_group_name",
                            "detail_desc"]


df_articles_sample.drop(redundo_catcols_to_drop, axis=1, inplace=True)

In [55]:
# Display basic column information for the initial Kaggle 'article' datatset - after Sampling of original dataset
df_articles_sample.head()

article_id  product_code  product_type_no  product_group_name  \
0   108775015        108775              253  Garment Upper body   
1   108775044        108775              253  Garment Upper body   
2   108775051        108775              253  Garment Upper body   
3   110065001        110065              306           Underwear   
4   110065002        110065              306           Underwear   

   graphical_appearance_no  colour_group_code  perceived_colour_value_id  \
0                  1010016                  9                          4   
1                  1010016                 10                          3   
2                  1010017                 11                          1   
3                  1010016                  9                          4   
4                  1010016                 10                          3   

   perceived_colour_master_id  department_no index_code  index_group_no  \
0                           5           1676          A               1   
1                           9           1676          A               1   
2                           9           1676          A               1   
3                           5           1339          B               1   
4                           9           1339          B               1   

   section_no  garment_group_no  
0          16              1002  
1          16              1002  
2          16              1002  
3          61              1017  
4          61              1017

#### Use One-Hot-Encoding on remaining categorical columns. <a class="tocSkip">

In [56]:
article_cols_to_encode =['product_group_name', 'index_code']
for column_name in article_cols_to_encode:
    df_articles_sample = encode_Categor_Features(df_articles_sample, column_name)

In [57]:
df_articles_sample.shape

(94127, 40)

In [58]:
# Display basic column information for sampled articles dataframe after encoding
df_articles_sample.head()

article_id  product_code  product_type_no  graphical_appearance_no  \
0   108775015        108775              253                  1010016   
1   108775044        108775              253                  1010016   
2   108775051        108775              253                  1010017   
3   110065001        110065              306                  1010016   
4   110065002        110065              306                  1010016   

   colour_group_code  perceived_colour_value_id  perceived_colour_master_id  \
0                  9                          4                           5   
1                 10                          3                           9   
2                 11                          1                           9   
3                  9                          4                           5   
4                 10                          3                           9   

   department_no  index_group_no  section_no  ...  A  B  C  D  F  G  H  I  J  \
0           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
1           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
2           1676               1          16  ...  1  0  0  0  0  0  0  0  0   
3           1339               1          61  ...  0  1  0  0  0  0  0  0  0   
4           1339               1          61  ...  0  1  0  0  0  0  0  0  0   

   S  
0  0  
1  0  
2  0  
3  0  
4  0  

[5 rows x 40 columns]

#### Merge Trxn/Customers Dataframe with Articles Dataframe (sample). <a class="tocSkip">

In [59]:
# Merge Articles atributes from the Articles dataset based on a join with the 'article_id' field
df_HM_CustTrxnArtcls = pd.merge(df_HM_CustTrxn, df_articles_sample, how="left", on=["article_id"])

In [60]:
# Confirm dimensions of Trxn (sample) dataframe AFTER merging with Articles data
df_HM_CustTrxnArtcls.shape

(6294126, 55)

In [61]:
# Display basic column information for the initial Kaggle 'transaction_train' datatset - AFTER merging with Articles data
df_HM_CustTrxnArtcls.head()

t_dat                                        customer_id  article_id  \
0 2019-10-18  27c7136517d21beeea675c41e1a57b23777d18ee8d870b...   733099003   
1 2019-04-10  5705316f00df6c607c50b3cdef5ce16246d8cb7aea83ac...   606395014   
2 2019-06-20  4173fc84e963778f9e24cb9e0e0586cb30d3233b3045ac...   665963007   
3 2019-07-02  1f8001b645e77313b15271e67a5010a6a3e277938c7fbc...   702890001   
4 2019-04-27  8839cf1ea0a62217e5c9a1b21b2a34e36cad0090040c3a...   375585027   

      price  sales_channel_id  txn_year  txn_mth  txn_day   age  ACTIVE  ...  \
0  0.010153                 1      2019       10       18  27.0       1  ...   
1  0.059305                 2      2019        4       10  30.0       1  ...   
2  0.006085                 1      2019        6       20  41.0       1  ...   
3  0.010153                 2      2019        7        2  27.0       1  ...   
4  0.007898                 2      2019        4       27  39.0       1  ...   

   A  B  C  D  F  G  H  I  J  S  
0  1  0  0  0  0  0  0  0  0  0  
1  0  0  0  0  1  0  0  0  0  0  
2  0  1  0  0  0  0  0  0  0  0  
3  0  1  0  0  0  0  0  0  0  0  
4  0  0  0  0  1  0  0  0  0  0  

[5 rows x 55 columns]

In [62]:
# Check for null values in the merged Trxn/Customer/Articles dataframe
[print('Any Null Values in the merged Trxn/Customer/Articles dataframe?','\n' ,df_HM_CustTrxnArtcls.isnull().any())]

Any Null Values in the merged Trxn/Customer/Articles dataframe? 
 t_dat                         False
customer_id                   False
article_id                    False
price                         False
sales_channel_id              False
txn_year                      False
txn_mth                       False
txn_day                       False
age                           False
ACTIVE                        False
LEFT CLUB                     False
PRE-CREATE                    False
Monthly                       False
NONE                          False
None                          False
Regularly                     False
product_code                  False
product_type_no               False
graphical_appearance_no       False
colour_group_code             False
perceived_colour_value_id     False
perceived_colour_master_id    False
department_no                 False
index_group_no                False
section_no                    False
garment_group_no              Fals

[None]

#### Final Data Preparation Steps

#### Remove Any Duplicates in Final Merged Dataframe. <a class="tocSkip">

In [63]:
# run a check to see if any duplicate entires exist in the merged Trxn/Customer/Articles dataframe
[print('Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe?','\n' 
                                                                       ,df_HM_CustTrxnArtcls.duplicated().sum())]

Numbers of duplicated rows in the final merged Trxn/Customer/Articles dataframe? 
 140997


[None]

In [64]:
# Drop duplicate rows in the merged Trxn/Customer/Articles dataframe
df_HM_CustTrxnArtcls = df_HM_CustTrxnArtcls.drop_duplicates()

In [65]:
# Re-check: Find duplicate rows across all columns
duplicateRows = df_HM_CustTrxnArtcls[df_HM_CustTrxnArtcls.duplicated()]

In [66]:
# This code will advise if duplicates are successfully removed.
if duplicateRows.shape[0] < 1:
    [print('Success! Duplicate Rows Removed from merged Trxn/Customer/Articles dataframe')]
else:
    [print('Warning! Duplicate Rows Remain in merged Trxn/Customer/Articles dataframe')] 

Success! Duplicate Rows Removed from merged Trxn/Customer/Articles dataframe


In [67]:
# Confirm dimensions of Trxn/Customer/Articles (sample) dataframe AFTER removing duplicate rows
[print('\nCurrent Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed','\n' ,
                                                                                           df_HM_CustTrxnArtcls.shape)]


Current Number of Rows/Columns in merged Trxn/Customer/Articles dataframe - duplicates removed 
 (6153129, 55)


[None]

In [68]:
### See if there is a way to check the integrity of the customer merging..

In [69]:
df_HM_CustTrxnArtcls.dtypes

t_dat                         datetime64[ns]
customer_id                           object
article_id                             int64
price                                float64
sales_channel_id                       int64
txn_year                               int32
txn_mth                                int32
txn_day                                int32
age                                  float64
ACTIVE                                 uint8
LEFT CLUB                              uint8
PRE-CREATE                             uint8
Monthly                                uint8
NONE                                   uint8
None                                   uint8
Regularly                              uint8
product_code                           int64
product_type_no                        int64
graphical_appearance_no                int64
colour_group_code                      int64
perceived_colour_value_id              int64
perceived_colour_master_id             int64
department

###### 

## Split Data: Train and Test Sets

#### Split Data into 70% for Training, 30% for Testing. <a class="tocSkip">

In [70]:
# Split the data into train and test set
# Parameter set to 30% for size of Test set
# Given that there is ordering in the dataset, the 'shuffle' parameter is set to 'True'
train, test = train_test_split(df_HM_CustTrxnArtcls, test_size=0.3, random_state=101, shuffle=True)

In [71]:
# Output Confirmation of data split
dataDescription = 'H&M Sample Data Split'
print("\n{} Training Set Shape : \n\t".format(dataDescription))
print(train.shape)
print("\n{} Test Set Shape : \n\t".format(dataDescription))
print(test.shape)


H&M Sample Data Split Training Set Shape : 
	
(4307190, 55)

H&M Sample Data Split Test Set Shape : 
	
(1845939, 55)


In [72]:
train.head()

t_dat                                        customer_id  \
5844114 2020-06-11  c64e5a41814ca552b2f2c52d9fc56cdad2b21c34ec634b...   
351448  2020-07-19  8aac726aacac06f49f17dd8a28caaf7feda6b01ba47864...   
1471607 2018-11-18  1e70882bca2e9e6b2e28f72a83b757be747db26a0c6037...   
5239265 2019-08-12  58ab24184c2858b037842c9d07d98d59820b9d8da4a192...   
6206624 2019-10-21  15046e5564f527f0cec360e1ba7edb9587c78727f9b14e...   

         article_id     price  sales_channel_id  txn_year  txn_mth  txn_day  \
5844114   770315005  0.016932                 1      2020        6       11   
351448    599580083  0.013542                 1      2020        7       19   
1471607   677049002  0.050831                 1      2018       11       18   
5239265   651687001  0.010153                 2      2019        8       12   
6206624   735840009  0.030508                 1      2019       10       21   

          age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
5844114  52.0       1  ...  1  0  0  0  0  0  0  0  0  0  
351448   59.0       1  ...  0  1  0  0  0  0  0  0  0  0  
1471607  32.0       1  ...  1  0  0  0  0  0  0  0  0  0  
5239265  23.0       0  ...  1  0  0  0  0  0  0  0  0  0  
6206624  51.0       1  ...  0  0  0  1  0  0  0  0  0  0  

[5 rows x 55 columns]

In [73]:
test.head()

t_dat                                        customer_id  \
2422330 2020-04-04  16d728161a35b05b2c56118306b09d528993dc43347030...   
4246960 2019-08-08  36f88702422db0b945cc7f8593734a1300fce7ad150da9...   
2977343 2020-01-24  40f37b9ac1a184f5934bc6f27124d5f47cf7095f84cbf9...   
1656531 2018-10-01  5c52622522e0935230baed21f271fc44c1eb896f86ef97...   
5498770 2020-07-16  be8e5f395f4f2d89957ba5be67c3a60a9f1c95e9d5862c...   

         article_id     price  sales_channel_id  txn_year  txn_mth  txn_day  \
2422330   804992021  0.020322                 2      2020        4        4   
4246960   762392001  0.016932                 2      2019        8        8   
2977343   659832001  0.005068                 2      2020        1       24   
1656531   574109014  0.021610                 2      2018       10        1   
5498770   816938002  0.022017                 1      2020        7       16   

          age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
2422330  19.0       1  ...  0  0  0  0  0  0  0  0  0  1  
4246960  50.0       1  ...  1  0  0  0  0  0  0  0  0  0  
2977343  22.0       1  ...  0  0  0  1  0  0  0  0  0  0  
1656531  28.0       1  ...  0  0  0  1  0  0  0  0  0  0  
5498770  35.0       1  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 55 columns]

In [74]:
custs = test.customer_id.unique()

In [75]:
custs.shape

(656318,)

In [76]:
custs

array(['16d728161a35b05b2c56118306b09d528993dc43347030ca249d323294545154',
       '36f88702422db0b945cc7f8593734a1300fce7ad150da9040b3aac802780e53b',
       '40f37b9ac1a184f5934bc6f27124d5f47cf7095f84cbf9d1fd4b1304ba3f5af7',
       ...,
       '52807c1558d777e318b76c56c7fdd2d0298c25615a7aa596047cd041967dbb11',
       'f50fccb846d4d92298716b1348eba5c87ec7ffe1dd66e1b6010be8d337d348a7',
       'f5683cf76a1fab11b9f38728d4ded4bed36c0f1dcebe697c58432d8ab9a804be'],
      dtype=object)

###### 

## Model Preparation

### Set Up the Scaling Algorithm

In [77]:
from sklearn.preprocessing import MinMaxScaler
# Normalizing numerical features so that each feature has a value between 0 and 1
scaler = MinMaxScaler()

In [78]:
# Normalizing numerical features so that each feature has mean 0 and variance 1
# feature_scaler = StandardScaler()

### Set Up the Columns to be Scaled

In [79]:
nn_cols =    ['price',
              'age',
              'product_code',
              'product_type_no',
              'graphical_appearance_no',
              'colour_group_code',
              'perceived_colour_value_id',
              'perceived_colour_master_id',
              'department_no',
              'index_group_no',
              'section_no',
              'garment_group_no']

### Scale the Training Dataset 

In [80]:
def scale_trainset_features(df, dtype, no_scale):
    for x in df.columns[df.dtypes == dtype]:
        if x != no_scale:
            df[x] = scaler.fit_transform(np.array(df[x]).reshape(-1,1))    

In [81]:
no_scale = 'article_id'
scale_dtypes = ['int64','int32','float64']
for data_types in scale_dtypes:
    scale_trainset_features(train, data_types, no_scale)

In [82]:
train.head()

t_dat                                        customer_id  \
5844114 2020-06-11  c64e5a41814ca552b2f2c52d9fc56cdad2b21c34ec634b...   
351448  2020-07-19  8aac726aacac06f49f17dd8a28caaf7feda6b01ba47864...   
1471607 2018-11-18  1e70882bca2e9e6b2e28f72a83b757be747db26a0c6037...   
5239265 2019-08-12  58ab24184c2858b037842c9d07d98d59820b9d8da4a192...   
6206624 2019-10-21  15046e5564f527f0cec360e1ba7edb9587c78727f9b14e...   

         article_id     price  sales_channel_id  txn_year   txn_mth   txn_day  \
5844114   770315005  0.028346               0.0       1.0  0.454545  0.333333   
351448    599580083  0.022614               0.0       1.0  0.545455  0.600000   
1471607   677049002  0.085669               0.0       0.0  0.909091  0.566667   
5239265   651687001  0.016882               1.0       0.5  0.636364  0.366667   
6206624   735840009  0.051304               0.0       0.5  0.818182  0.666667   

              age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
5844114  0.433735       1  ...  1  0  0  0  0  0  0  0  0  0  
351448   0.518072       1  ...  0  1  0  0  0  0  0  0  0  0  
1471607  0.192771       1  ...  1  0  0  0  0  0  0  0  0  0  
5239265  0.084337       0  ...  1  0  0  0  0  0  0  0  0  0  
6206624  0.421687       1  ...  0  0  0  1  0  0  0  0  0  0  

[5 rows x 55 columns]

In [83]:
test.head()

t_dat                                        customer_id  \
2422330 2020-04-04  16d728161a35b05b2c56118306b09d528993dc43347030...   
4246960 2019-08-08  36f88702422db0b945cc7f8593734a1300fce7ad150da9...   
2977343 2020-01-24  40f37b9ac1a184f5934bc6f27124d5f47cf7095f84cbf9...   
1656531 2018-10-01  5c52622522e0935230baed21f271fc44c1eb896f86ef97...   
5498770 2020-07-16  be8e5f395f4f2d89957ba5be67c3a60a9f1c95e9d5862c...   

         article_id     price  sales_channel_id  txn_year  txn_mth  txn_day  \
2422330   804992021  0.020322                 2      2020        4        4   
4246960   762392001  0.016932                 2      2019        8        8   
2977343   659832001  0.005068                 2      2020        1       24   
1656531   574109014  0.021610                 2      2018       10        1   
5498770   816938002  0.022017                 1      2020        7       16   

          age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
2422330  19.0       1  ...  0  0  0  0  0  0  0  0  0  1  
4246960  50.0       1  ...  1  0  0  0  0  0  0  0  0  0  
2977343  22.0       1  ...  0  0  0  1  0  0  0  0  0  0  
1656531  28.0       1  ...  0  0  0  1  0  0  0  0  0  0  
5498770  35.0       1  ...  1  0  0  0  0  0  0  0  0  0  

[5 rows x 55 columns]

###### 

###### 

###### 

###### 

## K Nearest Neighbour Operations

### Create Functions for NN Analysis

#### Set up function to Build Nearest Neighbours Model 

In [84]:
# Set up Unsupervised learner for implementing neighbour searches of Customer Transactions
# Set default parameters for NN algorithm - these can be overwritten during optimisationphase
def setup_NN_model(df_train, NN_metric = 'cosine', NN_algo = 'brute'):
       
    # Set up numerical data inputs for NearestNeighbors function
    df_cust_articles = df_train[nn_cols]
    
    # Set up parameters of NearestNeighbors function
    model_knn = NearestNeighbors(metric = NN_metric, algorithm = NN_algo)
    
    # Implement NearestNeighbors function on merged Trxn/Customer/Article data elements
    model_knn.fit(df_cust_articles)
    
    return model_knn, df_cust_articles

#### Set up function to  Nearest Neighbours (+ Articles bought) for a Customer 

In [85]:
# Use a function to return a list of Nearest Neighbours for a given customer.
# Return Neighbour Customer Ids and the articles they both
def get_CustSub_NeighbourIDs_articles(model_knn, df_cust_articles, df_train, NN_Neighbours):
     
    
    # Return index values for nearest neighbours - determined by 'Num_Neighbour' parameter
    distances, indices = model_knn.kneighbors(df_cust_articles.iloc[0].values.reshape(1, -1), 
                                              n_neighbors = NN_Neighbours)
     
    

    # Using the Index values returned by the Nearest neighbour model
    # Generate a list of Customer Ids and Article Ids for these neighbours
    for i in range(0, len(distances.flatten())):
        if i == 0:
            customer_list = []
            article_list  = []
        else:
            customer_list.append(df_train.loc[indices.flatten()[i],'customer_id'])
            article_list.append(df_train.loc[indices.flatten()[i],'article_id'])
        
    df_neighbourhood = pd.DataFrame({'customer_id': customer_list, 'article_id': article_list})
    del customer_list, article_list
    
    return df_neighbourhood  

#### Find Top Selling Products in 'Neighbourhood' for a Customer 

In [86]:
# Function to return Top 12 Articles Purchased by the Customer 'Neighbourhood' Or Actual Data on Purchases
def get_Top12_Articles_for_Cust(df, num_of_articles):
    
    #print('Display Total Cust Rec..')
    #display(df)
      
    # Finding number of times a particular item is purchased by customers in the neighbourhood
    temp = df.groupby(['article_id'])['customer_id'].agg('count').reset_index() 
    temp.columns = ['article_id','count']
    
    # Sort the articles by the number of time purchased by the 'neighbourhood' customers
    temp = temp.sort_values('count', ascending=False)
    
    # Select the Top 12 - if there are duplicate just select the first entry
    temp = temp.nlargest(num_of_articles,'count', keep='first')
    
    # Isolate the Article Ids
    temp = temp['article_id']
    
    # Convert to dataframe and reset index
    df_article_list = pd.DataFrame({'article_id': temp})
    df_article_list = df_article_list.reset_index(drop=True)
    
    
    # Return the list of Top 12 articles
    return df_article_list

#### Invoke Function to Build KNN Model  

In [87]:
# Reset Training dataframe index - remove old one; this allows for extraction of correct index records
# returned by the NearestNeighbors function
train.reset_index(drop=True,inplace=True)

In [88]:
# Call function to set up NN learner
hm_model_knn, df_cust_articles_nn = setup_NN_model(train, NN_metric = 'cosine', NN_algo = 'brute')

###### 

###### 

###### 

## Evaluate Model

#### The code segments below implement functions to extract predictions and evaluate results.  <a class="tocSkip">

### Define Functions to Extract Customer IDs and Article Predictions from Test Data

#### Define Function to read Customer IDs from Submission File

In [89]:
# Function to return Customer Ids from sample submission file in sequence
def get_next_SubmissionCustomer(idx):
           
    # Get Customer Id from Sample Sub
    cust_id       = df_sub_sample.iloc[idx,0]
    
    df_cust_id = df_sub_sample.head(idx+1)
    df_cust_id = df_cust_id.drop(['prediction'], axis = 1)
    
    
    
    # Return index of customer id in training dataframe and the customer id itself
    return df_cust_id, cust_id

#### Define Function to Obtain Customer Transactions

In [90]:
def find_latest_CustomerTrxn(rec):
    
    # Get Next Customer Id from Submission Table
    df_one_cust, cust_id = get_next_SubmissionCustomer(rec)

    
    # Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
    df_SubCust_Trxn = pd.merge(df_one_cust, df_trxns, how="left", on=["customer_id"])

    
    # On rare occassion that a customer has no transactions - create a dummy records to avoid a failure
    if df_SubCust_Trxn.shape[0] == 0:
        print('NO Transactions!')
        # Create dummy dataframe
        df_Dummy = pd.DataFrame([[cust_id,
                '2016-10-25',
                 651329001,
                 0.021593,
                 1]], columns=['customer_id','t_dat','article_id','price','sales_channel_id'])
        # Stack the DataFrames on top of each other
        df_SubCust_Trxn = pd.concat([df_SubCust_Trxn, df_Dummy], axis=0)
        df_SubCust_Trxn.reset_index(drop=True, inplace=True)
    
    
    # Use last transaction
    df_SubCust_Trxn1 = df_SubCust_Trxn.sort_values(by=['t_dat'], ascending=False).head(1)
    
    
    
    return df_SubCust_Trxn1, cust_id

#### Define Function to Obtain Customer Features

In [91]:
def find_latest_CustomerAttrs(df):

    
    # Merge Customer atributes from the Customer dataset based on a join with the 'customer_id' field
    df = pd.merge(df, df_customers, how="left", on=["customer_id"])
    
    # Filter out other Customer specific attributes apart from; age + fashion_news_frequency + club_member_status + postal code
    # There are very many missing values in the other Customer dataset columns, so these will be ignored
    df = df.iloc[:, [0, 1, 2, 3, 4, 7, 8, 9, 10]]

    
    
    return df

#### Define Function to Obtain Article Features

In [92]:
def find_latest_ArticleAttrs(df):
    
    # Merge Articles atributes from the Articles dataset based on a join with the 'article_id' field
    df = pd.merge(df, df_Articles, how="left", on=["article_id"])
     
    
    # Remove 'redundant' categorical columns from Articles (sample) dataframe
    # Consider a categorical column Redundant when is 'partnered' with a numerical code    
    df.drop(redundo_catcols_to_drop, axis=1, inplace=True)

    # Encode Categorical Values
    for column_name in article_cols_to_encode:
        df = encode_Categor_Features(df, column_name)
    
    
    return df    

#### Define Function to Obtain Article Predictions

In [93]:
def get_ArticlePredictions(knn_mdl, df, iNum_of_neighbours, TopN):
        
    # Call function to find nearest neighbour Customer Ids and Article Ids
    # Pass in index, parameters for Nearest Neighbour function, and the number of neighbours to retrun
    df_neighbours = get_CustSub_NeighbourIDs_articles(knn_mdl, df, train, iNum_of_neighbours)
    
    # Call function to return list of the most popular Articles in terms of the 
    # purchases made by the customers neighbourhood
    df_cust_preds = get_Top12_Articles_for_Cust(df_neighbours, TopN)

    
    return df_cust_preds    

#### Define Functions to Parse Article Predictions for Kaggle Submission

In [94]:
# Function to return Article Id Predictions in String format required for Kaggle competition
def get_str_Article_Predictions(df):
    
    article_pred = df['article_id'].values.tolist()
    
    article_pred = ['0' + str(article_id) for article_id in article_pred]
    article_pred_str =  ' '.join(article_pred)
    
    str_article_preds = article_pred_str
    
    return str_article_preds

In [95]:
# Function to Build up dataframe for submission
def build_up_submission_df(cust, str_preds, custid_str, last_col):
    
    lst = [[cust, str_preds]]    
    #df = pd.DataFrame(lst, columns =['customer_id', 'prediction'])
    df = pd.DataFrame(lst, columns =[custid_str, last_col])
    final_df =df
    
    
    return final_df

### Define Functions to Evaluate Model Predictions aginst Test Data

In [96]:
def evalute():
    print('Evaluate..')

In [97]:
eval_Cols = ['customer_id', 'prediction_accuracy']

#### Set up function to return latest Customer transactions in the Test Data 

In [98]:
test.sort_values('t_dat').groupby('customer_id').head(1)

t_dat                                        customer_id  \
1616213 2018-09-20  c1ce75cf6a97904139779ef31d6a3f3e200c898549357a...   
113468  2018-09-20  de71923e5a998dcc52709646309fb326878aed8a448539...   
5749305 2018-09-20  2696e69771b50e6902908647cc915544151554ca38c693...   
4102590 2018-09-20  19ce62097c4188f15888c9b8ef21e977195d6fcb510e4d...   
181297  2018-09-20  48fd2226ee8c854a688a94c4eed4cfaa8d620148661fa5...   
...            ...                                                ...   
6195926 2020-09-22  7b860822704b15a0e394d9f5cf8a2cc3a0b934824a0eda...   
5597919 2020-09-22  5e8fb4d457fdffc61e235328ba7e43a4139c94c5f9d52a...   
2641561 2020-09-22  c86896315044da9c01e6d8b7d5da2c4733cecfb7ccfb77...   
4937026 2020-09-22  cef3ccff7653679d44f4da7402f5db6ce91a21abd52507...   
2099750 2020-09-22  638f6161186115d63d1950e85c0ddadf5668af22c44157...   

         article_id     price  sales_channel_id  txn_year  txn_mth  txn_day  \
1616213   452511001  0.011847                 2      2018        9       20   
113468    622964001  0.010153                 2      2018        9       20   
5749305   652142002  0.016932                 1      2018        9       20   
4102590   573085019  0.033881                 2      2018        9       20   
181297    667491006  0.023915                 2      2018        9       20   
...             ...       ...               ...       ...      ...      ...   
6195926   918522001  0.042356                 1      2020        9       22   
5597919   849588007  0.050831                 2      2020        9       22   
2641561   919829001  0.042356                 2      2020        9       22   
4937026   803468002  0.025407                 2      2020        9       22   
2099750   556539003  0.005068                 1      2020        9       22   

          age  ACTIVE  ...  A  B  C  D  F  G  H  I  J  S  
1616213  22.0       1  ...  0  0  1  0  0  0  0  0  0  0  
113468   20.0       1  ...  0  0  0  1  0  0  0  0  0  0  
5749305  31.0       1  ...  0  0  0  1  0  0  0  0  0  0  
4102590  23.0       1  ...  1  0  0  0  0  0  0  0  0  0  
181297   19.0       1  ...  1  0  0  0  0  0  0  0  0  0  
...       ...     ...  ... .. .. .. .. .. .. .. .. .. ..  
6195926  19.0       1  ...  1  0  0  0  0  0  0  0  0  0  
5597919  26.0       1  ...  1  0  0  0  0  0  0  0  0  0  
2641561  31.0       1  ...  1  0  0  0  0  0  0  0  0  0  
4937026  28.0       1  ...  0  1  0  0  0  0  0  0  0  0  
2099750  26.0       1  ...  0  0  1  0  0  0  0  0  0  0  

[656318 rows x 55 columns]

In [99]:
def get_Latest_CustTxn_in_TestData():
    print('get_unique_CustIds_in_TestData')  
    
    df = test.sort_values('t_dat').groupby('customer_id').head(1)
    
    return df

#### Define Function to Return Evaluation Score

In [100]:
def return_Cust_Prediction_Score(cust_id, df_pred, df_actual, df_score):
    #print('Retrun Customer Prediction Score...')
    
    #df_match = df_pred.compare(df_actual, keep_equal=True)
    

    #"""Returns just the rows from the new dataframe that differ from the source dataframe"""
    merged_df = df_pred.merge(df_actual, indicator=True, how='outer')
    changed_rows_df = merged_df[merged_df['_merge'] == 'left_only']
    df_no_pred_match = changed_rows_df.drop('_merge', axis=1)
    
    #print('\ndf_no_pred_match...')
    #display(df_no_pred_match)
    
    scr_pct = (len(df_pred) - len(df_no_pred_match))/len(df_pred)
    
    lst = [[cust_id, scr_pct]]    
    df = pd.DataFrame(lst, columns = eval_Cols)
    
    # df_score = df
    
    # Stack the DataFrames on top of each other
    # df_Submission_Stack = pd.concat([df_Submission_Stack, df_NextCust_Preds], axis=0)
    df_score = pd.concat([df_score, df], axis=0)
    
    
    return df_score

#### Iterate Through Test Data and Generate Predictions

In [101]:
# Creating an empty Dataframe with column names only - will be filled with each predicton evaluation
df_ModelEvalResults_Stack = pd.DataFrame(columns= eval_Cols)

In [102]:
## Work Through Test Data to obtain prediction list of Articles
customer_records = 10000

df_Testdata_Customers_LastTrxn = get_Latest_CustTxn_in_TestData()
[print('Number of customers in Test Data is: {}\n'.format(len(df_Testdata_Customers_LastTrxn)))]

# Generate article predictions for each customer write to a new submission file
for cust_rec in range(0, customer_records):
    
    #print('cust_rec is...  {}\n'.format(df_Testdata_Customers_LastTrxn.iloc[cust_rec]))
       
    index_list = [cust_rec]
    df_CustRec = df_Testdata_Customers_LastTrxn.loc[df_Testdata_Customers_LastTrxn.index[index_list]]      
    df_CustTestNNVectors  = df_CustRec[nn_cols]
    
    
    C_Id = df_CustRec.iloc[0]['customer_id']
    
    df_CustRecId = df_CustRec[['customer_id']].copy()
        
    
    #print('Display: C_Id')
    #display(C_Id)
    
    #print('Display: df_CustRec')
    #display(df_CustRec)
    
    #print('Display: df_CustRecId')
    #display(df_CustRecId)
    #display(df_CustTestNNVectors)
    
    iNeighbours       = 550 # Number of Neighbours parameter for KNN model
    TopN              = 12  # Number of Artcile predictions to return
    
    df_TestCust_Preds     = get_ArticlePredictions(hm_model_knn, df_CustTestNNVectors, iNeighbours, TopN)
    
        
    # Create a dataframe of all transactions by the customer in the Test Data
    df_test = test.loc[test["customer_id"] == C_Id]
    #print('Display ...loc condition dataframe of all transactions by the customer in the Test Data')
    #display(df_test)
    
    
    df_TestCust_ActualPurchases = get_Top12_Articles_for_Cust(df_test, TopN)
    
    #display(df_TestCust_Preds)
    #display(df_TestCust_ActualPurchases)
    
    df_ModelEvalResults_Stack = return_Cust_Prediction_Score(df_CustRecId['customer_id'], 
                                                             df_TestCust_Preds, 
                                                             df_TestCust_ActualPurchases,
                                                             df_ModelEvalResults_Stack)
    
    

    

get_unique_CustIds_in_TestData
Number of customers in Test Data is: 656318



In [103]:
df_ModelEvalResults_Stack.reset_index(drop=True,inplace=True)
print('Show Results scores..')
display(df_ModelEvalResults_Stack)
display(df_ModelEvalResults_Stack.sort_values('prediction_accuracy', ascending=False))


Show Results scores..


customer_id  prediction_accuracy
0     1616213    c1ce75cf6a97904139779ef31d6a3f3e200...                  0.0
1     113468    de71923e5a998dcc52709646309fb326878a...                  0.0
2     5749305    2696e69771b50e6902908647cc915544151...                  0.0
3     4102590    19ce62097c4188f15888c9b8ef21e977195...                  0.0
4     181297    48fd2226ee8c854a688a94c4eed4cfaa8d62...                  0.0
...                                                 ...                  ...
9995  4770670    a1dc3abc4b71d4cc83c2ddfe2df5f8aebad...                  0.0
9996  5822271    509960a699ccb7d82e5e8fadbe50da15c8b...                  0.0
9997  1167617    9f5ee8b0a76cf732d5c26c9d4e120a71605...                  0.0
9998  1404213    097c477873a0a60f201fde02bf9f5341da8...                  0.0
9999  4675026    75ead485e489c02ea66a00f64ad07aab72a...                  0.0

[10000 rows x 2 columns]

customer_id  prediction_accuracy
166   3376263    907dd32a97c0b87cbf6056c0d3429112195...             0.083333
2534  2425436    bb7fbe2d33908ecd7a5d863a3ebe324ae54...             0.083333
546   1347940    166d52e135951c0b1120351897d7cd60b4a...             0.083333
8363  733254    8ce624c145ffa65141ef4ee82be7c9b153cc...             0.083333
257   4820504    d5097d5b9a4a182d9cee8e838659bfb14f3...             0.083333
...                                                 ...                  ...
3336  895946    4bbf960b9806666df12d71f8819c86739019...             0.000000
3337  1220249    289b3e675b1e6e903ce35ec9f01dc5c7a88...             0.000000
3338  2697010    cde26b12203531783553e55393dd3ca1f3d...             0.000000
3339  3595116    9ab1d0a3f026823ec1a2fd86d504b076583...             0.000000
9999  4675026    75ead485e489c02ea66a00f64ad07aab72a...             0.000000

[10000 rows x 2 columns]

In [117]:
df_teststack = df_ModelEvalResults_Stack.loc[df_ModelEvalResults_Stack["prediction_accuracy"] > 0]

In [119]:
df_teststack.shape

(17, 2)

###### 

###### 

## Build Submission

#### The solution file to have 1371980 prediction rows - use Sample Submission to construct  <a class="tocSkip">

### Read Sample Submission File

In [104]:
# Read in Full Sample Submission File
df_sub  = pd.read_csv('../input/sample_submission.csv', dtype={'customer_id': 'string'})

In [105]:
# Temp code to build submission framework based on a smaller fraction of the Sample Submission file
df_sub_sample = df_sub.sample(frac=0.001, replace=False)  

In [106]:
# Temp code to check dataframe matches file output
df_sub_sample.head()

customer_id  \
444811   5327f0431e309ea747e03451be659eedde5a6002459f3b...   
1186096  dd5008ed556a5f51828a7a71725a30d0d031a948c0282d...   
650102   7956644757611fcc5073813899132fb7c71d655b83aa1b...   
600738   7016ee6360096ffc0b82b782c59f1f48612c462daada7f...   
280018   3442f5fe61a6ecf35ec7fbe222515166b789f7303c7147...   

                                                prediction  
444811   0706016001 0706016002 0372860001 0610776002 07...  
1186096  0706016001 0706016002 0372860001 0610776002 07...  
650102   0706016001 0706016002 0372860001 0610776002 07...  
600738   0706016001 0706016002 0372860001 0610776002 07...  
280018   0706016001 0706016002 0372860001 0610776002 07...

#### Record record size of file for later validation.  <a class="tocSkip">

In [107]:
numCustomers = df_sub.shape[0]

### Iterate Through Submission File to Generate Article Predictions

In [108]:
# Creating an empty Dataframe with column names only - will be filled with each Customer and their set of predictions
df_Submission_Stack = pd.DataFrame(columns=['customer_id', 'prediction'])

In [109]:
# Set up loop based on Submission file size
customer_records = 0 # Skip this processing if zero


# Generate article predictions for each customer write to a new submission file
for cust_rec in range(0, customer_records):
    
    print('cust_rec is... ' + str(cust_rec) +'\n')
    
    df_SubCust_Trxn1, cust_id  = find_latest_CustomerTrxn(cust_rec)
    
    df_SubCust_Trxn1  = find_latest_CustomerAttrs(df_SubCust_Trxn1)
    
    df_SubCust_Trxn1  = find_latest_ArticleAttrs(df_SubCust_Trxn1)
    
    df_CustNNVectors  = df_SubCust_Trxn1[nn_cols]
    
    display(df_CustNNVectors)
        
    iNeighbours       = 550 # Number of Neighbours parameter for KNN model
    TopN              = 12  # Number of Artcile predictions to return
    df_cust_preds     = get_ArticlePredictions(hm_model_knn, df_CustNNVectors, iNeighbours, TopN)
    
    # Get a string output of Customer Article Predicitons - format as required for Kaggle competition
    str_articles_predictions = get_str_Article_Predictions(df_cust_preds)
    
    # Format Customer ID and list of Article predictions - add to dataframe to be used for CSV output
    df_NextCust_Preds = build_up_submission_df(cust_id, str_articles_predictions,'customer_id','prediction')
    
    
    # Stack the DataFrames on top of each other
    df_Submission_Stack = pd.concat([df_Submission_Stack, df_NextCust_Preds], axis=0)
     
    

### Write Predictions Submission File to CSV

#### SetUp Disk Directory for Submission File

In [110]:
# Set up folder for submission file with predictions
submission_path = '../submission'

In [111]:
# Create folder above project directory
os.makedirs(submission_path, exist_ok=True)  

#### Create Submission .CSV File

In [112]:
df_Submission_Stack = df_Submission_Stack.reset_index(drop=True)

In [113]:
# Write pre-test submission file to disk
df_Submission_Stack.to_csv(submission_path+'/submission_pretest.csv',index=False)

In [114]:
dfTempFinal  = pd.read_csv(submission_path+'/submission_pretest.csv')

[print("Matching the number of dfSub rows to the correct submission size. \nSubmission Row: {} v Expected Rows: {}"
                                                                                   .format(dfTempFinal.shape[0],numCustomers))]

Matching the number of dfSub rows to the correct submission size. 
Submission Row: 0 v Expected Rows: 1371980


[None]

In [115]:
# Check format of final submission file
dfTempFinal.head()

Empty DataFrame
Columns: [customer_id, prediction]
Index: []

In [116]:
# Write final submission file to disk
dfTempFinal.to_csv(submission_path+'/submission.csv',index=False)

###### 

###### 

## References

#### <a class="tocSkip">